analysis/time-analysis/time_analysis_main.py

```diff
@@ -1,98 +1,77 @@
+    :param min_sstubs_percentage: minimum percentage of bugs that must be part of interval
+    :return: shortest time interval in which at least the given percentage of time stamps is covered
     '''
-    print('')
-    print(' *** Printing min, max timestamps *** ')
-    filteredHashTimestamps = []
-    allDiffTimes = []
-    bucketDict = dict()
-    for y in filteredBucketHashList:
-        print('')
-        print('Bucket Hash: {}'.format(y))
-
-        for i, entry in enumerate(data):  # attempt to grab all timestamps where json has this hash value
-
-            hash = data[i]['bucketHash']
-            time = data[i]['testTime']
-            if hash == y:
-                time = datetime.strptime(time, "%b %d %Y %H:%M:%S").date()
-                # time = datetime.strptime(time, '%Y-%m-%dT%H:%M:%SZ').date()
-                filteredHashTimestamps.append(time)
-
-        minTime = min(filteredHashTimestamps)
-        maxTime = max(filteredHashTimestamps)
-        diffTime = maxTime - minTime
-        allDiffTimes.append(diffTime.days)
-
-        bucketDict[y] = {'timestamps': filteredHashTimestamps, 'diffTime': diffTime.days}
-
-        print('Entry Count: {}'.format(len(filteredHashTimestamps)))
-        print('Earliest Timestamp: {}'.format(minTime))
-        print('Latest Timestamp: {}'.format(maxTime))
-        print('Timestamp Diff (Days): {}'.format(diffTime.days))
-        print('Time Interval: {}'.format(get_interval(filteredHashTimestamps, minTime, maxTime)))
-        filteredHashTimestamps = []
 
-    print('')
-    print(bucketDict)
-    print("\n".join("{}\t{}".format(k, v) for k, v in bucketDict.items()))
-    return bucketDict
+    # calculate number of timestamps that must be within interval (ceil always rounds up)
+    min_timestamps_amount = int(math.ceil(len(timestamps_list) * min_sstubs_percentage))
+    sorted_timestamps = sorted(timestamps_list)
 
+    shortest_interval = 0
+    for i in range(len(timestamps_list)):
+        first_timestamp_index = i
+        last_timestamp_index = first_timestamp_index + min_timestamps_amount
 
-def get_unique(bucketHashList):
-    '''
-    Get unique hashes from json
-    '''
-    unique = []
+        if last_timestamp_index >= len(sorted_timestamps):
+            break
+
+        diffTime = sorted_timestamps[last_timestamp_index] - sorted_timestamps[first_timestamp_index]
+        if i == 0 or diffTime.days < shortest_interval:
+            shortest_interval = diffTime.days
 
-    for hash in bucketHashList:
-        if hash in unique:
-            continue
-        else:
-            unique.append(hash)
-    return unique
+    return shortest_interval
 
 
-'''
-Loop through and grab bucket hashes.
-'''
+def get_projects_to_bucket_details():
+    data = get_grouped_sstubs()
+    projects_to_buckets_details = dict()
 
-with open('testTimestamps.json') as json_file:
-    data = json.load(json_file)
+    for project_name, bucket in data.items():
```

Comment:
best rename `bucket` to `buckets` or even better `buckets_to_sstubs`, as it's not necessarily a single bucket but rather a dictionary of buckets and the related sstubs